pkg/mutation/system.go

```diff
@@ -100,12 +108,60 @@ func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) (b
 			if cmp.Equal(original, obj) {
 				return false, fmt.Errorf("oscillating mutation for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
 			}
+			logAppliedMutations(original, allAppliedMutations)
 			return true, nil
 		}
+		allAppliedMutations = append(allAppliedMutations, appliedMutations)
 	}
 	return false, fmt.Errorf("mutation not converging for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
 }
 
+func logAppliedMutations(obj *unstructured.Unstructured, allAppliedMutations [][]types.Mutator) {
+	iterations := []interface{}{}
+	for i, appliedMustations := range allAppliedMutations {
+		if len(appliedMustations) == 0 {
+			continue
+		}
+		var appliedMutationsText strings.Builder
+		mutationSeparator := ""
+		for _, mutator := range appliedMustations {
+			fmt.Fprintf(&appliedMutationsText, "%s%s", mutationSeparator, mutatorKey(mutator))
+			mutationSeparator = ", "
+		}
+		iterations = append(append(iterations, fmt.Sprintf("iteration_%d", i)), appliedMutationsText.String())
+	}
+	if len(iterations) > 0 {
+		logDetails := []interface{}{}
+		logDetails = append(append(logDetails, logging.EventType), logging.MutationApplied)
+		logDetails = append(append(logDetails, logging.ResourceGroup), obj.GroupVersionKind().Group)
+		logDetails = append(append(logDetails, logging.ResourceKind), obj.GroupVersionKind().Kind)
+		logDetails = append(append(logDetails, logging.ResourceAPIVersion), obj.GroupVersionKind().Version)
+		logDetails = append(append(logDetails, logging.ResourceNamespace), obj.GetNamespace())
+		logDetails = append(append(logDetails, logging.ResourceName), obj.GetName())
+		logDetails = append(logDetails, iterations...)
+		log.Info("Mutation applied.", logDetails...)
+	}
+}
+
+func mutatorKey(mutator types.Mutator) string {
```

Comment:
I'm thinking we could make this a `String()` method for the object returned by `mutator.ID()` so that it implements the Stringer interface:

https://tour.golang.org/methods/17

Then we can just provide `mutator.ID()` directly to `Fprintf()`

Reply:
Added the String() to Mutator.